### ISSUE TITLE
Rotation speed

### ISSUE THREAD

**Issue by cxong** [cxong/cdogs-sdl#370]
In C-Dogs, when players move around they change direction immediately - press left and the player faces left, press right and the player faces right - but what if there's a process where the player turns to face the new direction, going through the intermediate ones - left, down/left, down, down/right, right?

This represents aesthetic and gameplay changes.

If the change is limited to aesthetics (i.e. the player isn't actually going through all the directions, and can shoot in the new direction immediately), then:

- It looks nice (maybe; let's try it out)
- We'll need to add a separate looking-at direction; this can be managed client-side

But things get interesting if we also change the gameplay. As always, gameplay changes as drastic as this needs to be configurable.

To make this really cool, the shooting direction should be refactored into angles, instead of just the 4 directions.

**On bots only**

C-Dogs bots can be artificially difficult because they can instantly turn around to face the player. The player can't really use tactics like encircling to surprise the bot. If we add a delay for bot turning, we can make the game more tactically interesting.

On the other hand, we'll need to modify the existing AI code so this change doesn't make some bots too ineffective. To be specific, bots that use very slow firing guns like the sniper rifle. One solution would be:

- Instead of firing straight away, bots set a "want to shoot in direction" state, and won't fire until they are facing that direction
- The rotation speed is tied to the delay parameter - they should be guaranteed to be able to face any direction within the delay interval
- Whilst in the "want to shoot in direction" state, they won't do anything apart from rotating
  - In terms of commands, they'll execute a move command at the start, and then wait until the game updates their direction to the desired one.
  - This means for very high delay AIs, they'll be standing still
- At the end of the state, i.e. when they are facing the desired direction, fire

One edge case is if their delay is higher than their weapon firing rate. For guns like the flamer or machine gun, it would be nice to have them firing as they are turning around. One heuristic would be:

- Whilst in the "want to shoot in direction" state, as long as their gun's firing delay is less than the time to the end of the delay, fire

**On weapons**

If this property were to be applied to players, it's best that it's done per weapon. Guns like sniper rifles and shotguns should have instant rotation speed, but machine guns and flamers might benefit from rotation. This also makes those weapons cooler by allowing them to "sweep".

This also leads to the possibility of using low rotation speed as a balancing mechanism; maybe a heavy machine gun that can't be turned around quickly.

To get the best of both worlds, rotation should be instant as long as the gun isn't firing.

So in summary:

- [ ] Add separate variable for looking at direction
- [ ] Add config for gameplay direction changing
- [ ] Refactor direction to be based on angle
- [ ] Modify AI code to use rotation speed, with rotation state and fire delay
- [ ] Add rotation speed property to guns
- [ ] Implement rotation speed, while fire is held down

**Comment by cxong** [comment]
One idea: don't use rotation if the player is switching to a non-adjacent direction, e.g. left to right, but do rotate if they are adjacent, e.g. left-upleft-up-upright-right.
